config/github/src/main/kotlin/GitHubConfigFileCache.kt

```diff
@@ -0,0 +1,168 @@
+
+            val needRetry = if (!dataFile.isFile) {
+                logger.info(
+                    "File '{}' at revision '{}' not found in cache, downloading it to '{}'.",
+                    path,
+                    revision,
+                    dataFile
+                )
+                downloadFile(dataFile, load)
+            } else {
+                false
+            }
+
+            if (needRetry) {
+                getOrPutFile(revision, path, load)
+            } else {
+                dataFile.inputStream().waitForReadLock()
+            }
+        }
+
+    /**
+     * Download a file using the given [load] function and write it into the cache as [dataFile]. Return a flag
+     * whether this was successful. A return value of *false* means that the same file is currently downloaded by
+     * a different process. Then the operation needs to be retried.
+     */
+    private suspend fun downloadFile(dataFile: File, load: suspend () -> ByteReadChannel): Boolean =
+        withContext(Dispatchers.IO) {
+            dataFile.parentFile.mkdirs()
+
+            RandomAccessFile(dataFile, "rw").use { file ->
+                file.channel.use { writeChannel ->
+                    val lock = runCatching { writeChannel.tryLock() }.getOrNull()
+
+                    // If no write lock can be acquired, the file is currently downloaded by another process. In this
+                    // case, simply retry the operation.
+                    lock?.use {
+                        if (dataFile.length() <= 0L) {
+                            val readChannel = load()
+
+                            while (!readChannel.isClosedForRead) {
+                                val packet = readChannel.readRemaining()
+                                while (!packet.isEmpty) {
+                                    val data = packet.readByteBuffer()
+                                    writeChannel.write(data)
+                                }
+                            }
+                        } else {
+                            logger.info("File '{}' already exists, skipping download.", dataFile)
+                        }
+                        false
+                    } ?: true
+                }
+            }
+        }
+
+    /**
+     * Wait until a read lock is available for the given [FileInputStream]. This is needed to ensure that a file has
+     * been fully downloaded until a stream to it is returned from the cache.
+     */
+    private suspend fun FileInputStream.waitForReadLock(): FileInputStream =
+        withContext(Dispatchers.IO) {
+            var lock: FileLock?
+
+            do {
+                lock = runCatching {
+                    this@waitForReadLock.channel.tryLock(0, Long.MAX_VALUE, true)
+                }.getOrNull()
+
+                if (lock == null) {
+                    delay(lockCheckInterval)
+                }
+            } while (lock == null)
+
+            lock.close()
```

Comment:
As `Filelock` implements `AutoCloseable`, can you use `.use{}` here ?

Reply:
This does not play well with the loop.